File: eval-dotnet/src/Eval/Declaration.cs
using System.Text.Json.Serialization;

namespace Eval;

/// <summary>What a run said it was doing, before it did it.</summary>
/// <remarks>
/// <para>
/// A declaration is a closed object even where the acceptance predicate is
/// open: whether it is <i>complete</i> is mechanically checkable, whatever the
/// quality of the work that follows. Prediction before operation, applied to
/// the worker.
/// </para>
/// <para>
/// Made before the act or not at all. A declaration written afterwards
/// describes what happened, which is a summary and not a prediction — and a
/// summary cannot be contradicted by the run it summarises.
/// </para>
/// </remarks>
/// <param name="Decision">The decision being resolved, in the worker's own words.</param>
/// <param name="Ground">The ground it says it needs. Named elements, not prose.</param>
/// <param name="Tolerance">
/// The declared bound on outcome-relevant variation. The arrangement's, not the
/// worker's — a worker setting its own tolerance is deciding how wrong it is
/// allowed to be.
/// </param>
/// <param name="Assurance">
/// The assurance level the act is performed at. Also the arrangement's: it is
/// derived from the worst credible outcome by whoever bears that outcome, and a
/// worker declaring it would be pricing a consequence it does not carry.
/// </param>
public sealed record Declaration(
    [property: JsonPropertyName("decision")] string Decision,
    [property: JsonPropertyName("ground")] IReadOnlyList<string> Ground,
    [property: JsonPropertyName("tolerance")] string? Tolerance = null,
    [property: JsonPropertyName("assurance")] string? Assurance = null)
{
    /// <summary>The fields this declaration is missing to be complete.</summary>
    public IReadOnlyList<string> Missing()
    {
        var missing = new List<string>();
        if (string.IsNullOrWhiteSpace(Decision))
        {
            missing.Add("decision");
        }
        if (Ground.Count is 0)
        {
            missing.Add("ground");
        }
        if (string.IsNullOrWhiteSpace(Tolerance))
        {
            missing.Add("tolerance");
        }
        if (string.IsNullOrWhiteSpace(Assurance))
        {
            missing.Add("assurance");
        }
        return missing;
    }

    /// <summary>Add what the arrangement — not the worker — fixed.</summary>
    public Declaration BoundedBy(string tolerance, string assurance) =>
        this with { Tolerance = tolerance, Assurance = assurance };
}

/// <summary>A claim the act made, and the declared ground it rests on.</summary>
/// <remarks>
/// A claim with no ground is an escape candidate: the worker resolved something
/// on a basis it did not declare, or on none.
///
/// <b>An attribution is the worker's own account of itself</b> and can be
/// confabulated. It is worth checking because it is cheap and because its
/// falsifier is sharp — attribution passing on acts where behaviour shows
/// dependence on undeclared ground means the attribution is decorative.
/// </remarks>
public sealed record Attribution(
    [property: JsonPropertyName("claim")] string Claim,
    [property: JsonPropertyName("ground")] string? Ground = null);

File: eval-core/src/declaration.rs
//! What a run said it was doing, before it did it.
//!
//! A declaration is a closed object even where the acceptance predicate is
//! open: whether it is *complete* is mechanically checkable, whatever the
//! quality of the work that follows. Prediction before operation, applied to
//! the worker.
//!
//! Made before the act or not at all. A declaration written afterwards
//! describes what happened, which is a summary and not a prediction — and a
//! summary cannot be contradicted by the run it summarises.

use serde::{Deserialize, Serialize};

/// What a worker declared before acting.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Declaration {
    /// The decision being resolved, in the worker's own words.
    pub decision: String,
    /// The ground it says it needs. Named elements, not prose.
    #[serde(default)]
    pub ground: Vec<String>,
    /// The declared bound on outcome-relevant variation.
    ///
    /// The arrangement's, not the worker's. A worker that set its own tolerance
    /// would be deciding how wrong it is allowed to be.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tolerance: Option<String>,
    /// The assurance level the act is being performed at.
    ///
    /// Also the arrangement's. It is derived from the worst credible outcome by
    /// whoever bears that outcome; a worker declaring it would be pricing a
    /// consequence it does not carry.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub assurance: Option<String>,
}

impl Declaration {
    /// A declaration of what is being resolved, over named ground.
    pub fn new(decision: impl Into<String>, ground: Vec<String>) -> Self {
        Self { decision: decision.into(), ground, tolerance: None, assurance: None }
    }

    /// Add what the arrangement — not the worker — fixed.
    pub fn bounded_by(
        mut self,
        tolerance: impl Into<String>,
        assurance: impl Into<String>,
    ) -> Self {
        self.tolerance = Some(tolerance.into());
        self.assurance = Some(assurance.into());
        self
    }

    /// The fields a declaration must carry to be complete.
    pub fn missing(&self) -> Vec<&'static str> {
        let mut missing = Vec::new();
        if self.decision.trim().is_empty() {
            missing.push("decision");
        }
        if self.ground.is_empty() {
            missing.push("ground");
        }
        if self.tolerance.as_deref().map(str::trim).unwrap_or_default().is_empty() {
            missing.push("tolerance");
        }
        if self.assurance.as_deref().map(str::trim).unwrap_or_default().is_empty() {
            missing.push("assurance");
        }
        missing
    }
}

/// A claim the act made, and the declared ground it rests on.
///
/// A claim with no ground is an escape candidate: the worker resolved
/// something on a basis it did not declare, or on none.
///
/// **An attribution is the worker's own account of itself** and can be
/// confabulated. It is worth checking because it is cheap and because its
/// falsifier is sharp — attribution passing on acts where behaviour shows
/// dependence on undeclared ground means the attribution is decorative.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Attribution {
    /// The claim, as the act stated it.
    pub claim: String,
    /// The declared ground element it rests on, where one was named.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ground: Option<String>,
}

impl Attribution {
    /// A claim resting on a named ground element.
    pub fn to(claim: impl Into<String>, ground: impl Into<String>) -> Self {
        Self { claim: claim.into(), ground: Some(ground.into()) }
    }

    /// A claim resting on nothing the worker named.
    pub fn unattributed(claim: impl Into<String>) -> Self {
        Self { claim: claim.into(), ground: None }
    }
}

#[path = "declaration_tests.rs"]
#[cfg(test)]
mod tests;
